### PR TITLE
Use proper pid namespace

### PR DESCRIPTION
**Pull request by an3l**
#### Problem:
==============
By testing `pgrep` with `--ns` option,
introduced with MDEV-21331, commit https://github.com/MariaDB/server/commit/fb7c1b9415c9a8b0dc2e86ae44f0e7a2634e5d7e,
I noted that:
a) `--ns`  cannot use more than single PID.
b) `--ns` is returning the processes of the namespace to which supplied PID belongs to.
So by that sense command `pgrep -x --ns $$ mysqld` will always return an error and skip
checking of the existing PID of the server.

#### Solution:
==============
    Suggested solution is to add `--nslist pid`, since `--ns` needs to know in which namespace type it should look for.
    See `pgrep --help` for different namespace types.
    Note also that this works *only* if script is run as a `root` (we have that case here).

Current PR is a part of:
1. MDEV-21331: sync preinst and postrm script
2. MDEV-15718: check for exact mysqld process

This commit:
a) fixes https://github.com/MariaDB/server/commit/fb7c1b9415c9a8b0dc2e86ae44f0e7a2634e5d7e
b) Closes PR https://github.com/MariaDB/server/pull/2068 (obsolete)
c) Closes PR https://github.com/MariaDB/server/pull/2069 (obsolete)

See more in JIRA comment of MDEV-15718.
Thanks Faustin Lammler <faustin@mariadb.org> for testing and verifying

Reviewed by: <>
## How can this PR be tested?

Please take a look into https://github.com/MariaDB/mariadb.org-tools/tree/master/anel/shell_scripts/pid_namespace

### CODE REVIEW

**Comment by ottok** [comment]
I don't fully understand the commit message body. Can't review quickly now.

If this PR is correct, then it would render https://github.com/MariaDB/server/pull/2069 and https://github.com/MariaDB/server/pull/2068 obsolete

**Comment by an3l** [comment]
Have updated the commit message for quick peek.

**Comment by vuvova** [comment]
I don't understand the commit message either. a) and b) are correct.
I don't see why "command `pgrep -x --ns $$ mysqld` will always return an error"

**Comment by an3l** [comment]
I guess because `--ns` takes single PID (and we are giving 2) ?
Here is the test https://github.com/MariaDB/mariadb.org-tools/blob/master/anel/shell_scripts/pid_namespace/testpidnamespace.sh#L42
and here is the result https://github.com/MariaDB/mariadb.org-tools/blob/master/anel/shell_scripts/pid_namespace/results.txt#L43
No output of command (also in the script are other examples).

**Comment by fauust** [comment]
To be tested but the following command seems to work:
```console
pgrep -x --nslist pid --ns $$ mysqld
<PID-LOCAL->

pgrep -x --nslist pid mysqld
<PID-LOCAL>
<PID-CONTAINER>
```

**Comment by vuvova** [comment]
no, `-x` is definitely needed. you don't know what else is running in this container

**Comment by an3l** [comment]
Hi @vuvova , @ottok 
TL;DR
I would still vote to go with my suggested patch in commit, because
1. It works for `mariadbd|mysqld`
2. I don't have crystal understanding of *why*  `pgrep --nslist` **why** works with/without `-x` for `mysqld` processes
3. If 2. is used we will need duplicate the code for `mariadbd` (10.5+) when needed (`pgrep` cannot apply logical or operator). With suggested patch I know exactly what I'm doing with bit more code for all releases.

However, let me explain if you decide to go with 2.  option (with `pgrep --nslist`) why I vote to go without `-x` option, what my current understanding is:
- To see the pid namespaces :
```bash
$ ps -eo pidns,pid,args | grep bash
4026531836  10328 bash
```
`4026531836 ` is `root namespace` on *nix.
- To see `mysqld` namespaces:
```bash
$ ps -eo pidns,pid,args|grep mysqld
         -  12066 mysqld                   container
         -  17006 /usr/sbin/mysqld    host
```
We have 2 processes with same name, different PID and **without namespace** (unnamed namespace). Reason that is unnamed is because is not run as root:
```bash
$ sudo ps -eo pidns,pid,args|grep mysqld
4026532622  12066 mysqld
4026531836  17006 /usr/sbin/mysqld
```
As can be seen `root` namespace `4026531836` is different than container namespace `4026532622`
Alternative  to find about container namespace run:
```bash
$ sudo ls -la /proc/12066/ns/|grep pid
lrwxrwxrwx 1 999 docker 0 Mar 31 00:50 pid -> pid:[4026532622]
lrwxrwxrwx 1 999 docker 0 Mar 31 04:04 pid_for_children -> pid:[4026532622]
```
- To get the container PID using `pgrep`, I read like this: "*Get from PID namespace type ALL process from namespace to which  12066 belongs to*" - since there is single container and single process, it will be single output
```bash
$ sudo pgrep --nslist pid --ns 12066
12066
```
- To start new process run:
```bash
$ docker exec -it mysql-cont bash
root@81a9aac16ca6:/#
# From other terminal
$ sudo pgrep --nslist pid --ns 12066
12066
19135
```
Now we have new PID listed `bash` (19135). We can get list of all processes with other process from the same namespace:
```bash
// The same is obtained with other process
$ sudo pgrep --nslist pid --ns 19135
12066
19135
```
- To get the `mysqld` in specific process - here is where things get weird:
```bash
$ sudo pgrep --nslist pid --ns 12066 12066 // this doesn't work
$ sudo pgrep --nslist pid --ns 19135 12066 // this doesn't work

$ sudo pgrep --nslist pid --ns 12066 mysqld // this works by getting the `name` without -x (it can be single mysqld only)
12066 
$ sudo pgrep --nslist pid --ns 19135 mysqld // this works too
12066
```
So I understand this: "**From namespace of PID (19135) find `mysqld` process by name  (not by number), so no need to use -x**". I read [pgrep source code](https://github.com/thlorenz/procps/blob/master/deps/procps/pgrep.c#L505) and still don't see why such behavior.

- Not related but good to know: each new container will have isolated namespace
```bash
$ docker container run --rm --name my2 -e MYSQL_ROOT_PASSWORD=secret -d mysql
$ ps -eo pidns,pid,args|grep mysqld
         -  12066 mysqld
         -  17006 /usr/sbin/mysqld
         -  18603 mysqld
```
Now we got the new container PID 18603, that has its own namespace `4026533257`
```bash
$ sudo ps -eo pidns,pid,args|grep mysqld
4026532622  12066 mysqld
4026531836  17006 /usr/sbin/mysqld
4026533257  18603 mysqld
```

**Comment by an3l** [comment]
Ok. Just to give that example too as a name `my_mysqld` :
```bash
$ ps aux | grep -E "mysqld|mariadbd"
mysql      6256  0.0  1.0 1808412 82140 ?       Ssl  06:31   0:00 /usr/sbin/mysqld
999        9135  0.0  0.9 1475580 77532 ?       Ssl  06:53   0:00 mariadbd
anel       9503 12.6  0.0   4516   864 pts/1    S+   06:58   0:13 ./my_mysqld

$ ps -eo pidns,pid,args|grep -E "mysqld|mariadbd"
         -   6256 /usr/sbin/mysqld
         -   9135 mariadbd
4026531836   9503 ./my_mysqld

$ sudo pgrep  --nslist pid --ns $$ "mysqld|mariadbd"
6256
9503

$ sudo pgrep  -x --nslist pid --ns $$ "mysqld|mariadbd"
6256
```

Patch updated. Thanks Faustin.

File: debian/mariadb-server-10.2.preinst [hunk 1]
```diff
@@ -22,10 +22,11 @@ mysql_upgradedir=/var/lib/mysql-upgrade
...
 stop_server() {
     if [ ! -x /etc/init.d/mysql ]; then return; fi
-
-    # Return immediately if there are no mysql processes running
+    # Return immediately if there are no mysql processes running on a host
+    # (leave containerized processes with the same name in other namespaces)
     # as there is no point in trying to shutdown in that case.
-    if ! pgrep --ns $$ mysqld > /dev/null; then return; fi
+    # Compatibility with versions that ran 'mariadbd'
+    if ! pgrep -x --nslist --ns $$ "mysqld|mariadbd" > /dev/null; then return; fi
```
Comment (fauust) [inline debian/mariadb-server-10.2.preinst:29]:
```diff
- if ! pgrep -x --nslist --ns $$ "mysqld|mariadbd" > /dev/null; then return; fi
+ if ! pgrep -x --nslist pid --ns $$ "mysqld|mariadbd" > /dev/null; then return; fi
```

**Comment by an3l** [comment]
@ottok can you please review ?

**Comment by ottok** [comment]
> @ottok can you please review ?

Sure, I added myself to reviewers now. I don't have time to review this right away, but I have it on my todo now (at https://github.com/pulls/review-requested).

**Review by ottok — APPROVED** [review]
OK, so we have this in the '.preinst` file since MariaDB 10.2, and in both `.preinst` and '.postinst´ since MariaDB 10.5. The check should definitely have `-x`. The question is about how to check namespace. [Man page of pgrep](https://manpages.debian.org/bullseye/procps/pgrep.1.en.html) says:

> --ns pid
> Match processes that belong to the same namespaces. Required to run as root to match processes from other users. See --nslist for how to limit which namespaces to match.
> --nslist name,...
> Match only the provided namespaces. Available namespaces: ipc, mnt, net, pid, user,uts.

Seems this is the correct solution now.

**Comment by ottok** [comment]
Tested on https://salsa.debian.org/mariadb-team/mariadb-server/-/pipelines/373380 - no regressions due to this! (failures on CI pipeline are due to other issues)

**Comment by an3l** [comment]
Thanks @ottok.
FYI [MDEV-28468](https://jira.mariadb.org/browse/MDEV-28468)

**Comment by ottok** [comment]
This was also applied in official Debian packaging in https://salsa.debian.org/mariadb-team/mariadb-server/-/commit/cbde64aadf576609c3d708838db37a6791589523
